ssl/ech/ech_internal.c

```diff
@@ -160,4 +214,1194 @@ int ossl_ech_conn_init(SSL_CONNECTION *s, SSL_CTX *ctx,
+        acp = s->s3.server_random + SSL3_RANDOM_SIZE - OSSL_ECH_SIGNAL_LEN;
+        memcpy(acbuf, acp, OSSL_ECH_SIGNAL_LEN);
+        return 1;
+    } else {
+        if (!PACKET_buf_init(&pkt, shbuf, shlen)
+            || !PACKET_get_net_2(&pkt, &pi_tmp)
+            || !PACKET_get_bytes(&pkt, &pp_tmp, SSL3_RANDOM_SIZE)
+            || !PACKET_get_1(&pkt, &pi_tmp) /* sessid len */
+            || !PACKET_get_bytes(&pkt, &pp_tmp, pi_tmp) /* sessid */
+            || !PACKET_get_net_2(&pkt, &pi_tmp) /* ciphersuite */
+            || !PACKET_get_1(&pkt, &pi_tmp) /* compression */
+            || !PACKET_get_net_2(&pkt, &pi_tmp)) /* len(extensions) */
+            return 0;
+        while (PACKET_remaining(&pkt) > 0 && done < 1) {
+            if (!PACKET_get_net_2(&pkt, &etype)
+                || !PACKET_get_net_2(&pkt, &elen))
+                return 0;
+            if (etype == TLSEXT_TYPE_ech) {
+                if (elen != OSSL_ECH_SIGNAL_LEN
+                    || !PACKET_get_bytes(&pkt, &acp, elen))
+                    return 0;
+                memcpy(acbuf, acp, elen);
+                done++;
+            } else {
+                if (!PACKET_get_bytes(&pkt, &pp_tmp, elen))
+                    return 0;
+            }
+        }
+        return done;
+    }
+    return 0;
+}
+
+/*
+ * make up a buffer to use to reset transcript
+ * for_hrr is 1 if we've just seen HRR, 0 otherwise
+ * shbuf is the output buffer
+ * shlen is the length of that buffer
+ * tbuf is the output buffer
+ * tlen is the length of that buffer
+ * chend returns the offset of the end of the last CH in the buffer
+ * fixedshbuf_len returns the fixed up length of the SH
+ * return 1 for good, 0 otherwise
+ */
+int ossl_ech_make_transcript_buffer(SSL_CONNECTION *s, int for_hrr,
+                                    const unsigned char *shbuf, size_t shlen,
+                                    unsigned char **tbuf, size_t *tlen,
+                                    size_t *chend, size_t *fixedshbuf_len)
+{
+    unsigned char *fixedshbuf = NULL, *hashin = NULL, hashval[EVP_MAX_MD_SIZE];
+    unsigned int hashlen = 0, hashin_len = 0;
+    EVP_MD_CTX *ctx = NULL;
+    EVP_MD *md = NULL;
+    WPACKET tpkt = { 0 }, shpkt = { 0 };
+    BUF_MEM *tpkt_mem = NULL, *shpkt_mem = NULL;
+
+    /*
+     * SH preamble has bad length at this point on server
+     * and is missing on client so we'll fix
+     */
+    if ((shpkt_mem = BUF_MEM_new()) == NULL
+        || !WPACKET_init(&shpkt, shpkt_mem)) {
+        BUF_MEM_free(shpkt_mem);
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    if (!WPACKET_put_bytes_u8(&shpkt, SSL3_MT_SERVER_HELLO)
+        || (s->server == 1 && !WPACKET_put_bytes_u24(&shpkt, shlen - 4))
+        || (s->server == 1 && !WPACKET_memcpy(&shpkt, shbuf + 4, shlen -4))
+        || (s->server == 0 && !WPACKET_put_bytes_u24(&shpkt, shlen))
+        || (s->server == 0 && !WPACKET_memcpy(&shpkt, shbuf, shlen))
```

Comment:
These four lines do something different depending on whether we're called server side or client side. But I don't see how we can end up here on the server side. This function only seems to be called by either `tls_process_server_hello` (which is a client side function), or from `ossl_ech_calc_confirm`, which itself is only called from `tls_process_server_hello`...so all callers I can see are client side. Can we lose the server side logic here?

Reply:
OBE - see below